src/lib.rs

```diff
@@ -640,6 +645,113 @@ impl fmt::Debug for StructuredExtendedInformation {
     }
 }
 
+#[derive(Copy,Clone)]
+pub struct ProcessorExtendedState {
+    eax: u32,
+    ebx: u32,
+    ecx: u32,
+}
+
+impl ProcessorExtendedState {
+    fn new() -> ProcessorExtendedState {
+        let (a, b, c, _) = cpuid(RequestType::ProcessorExtendedState);
+        ProcessorExtendedState { eax: a, ebx: b, ecx: c }
+    }
+
+    bit!(eax, {
+        0 => x87_state,
+        1 => sse_state,
+        2 => avx_state,
+        // 3-4 mpx_state
+        // 5-7 avx_512_state
+        8 => ia32_xss,
+        9 => pkru_state
+        // 10-31 - reserved
+    });
+
+    pub fn mpx_state(self) -> u32 {
+        bits_of(self.eax, 3, 4)
+    }
+
+    pub fn avx_512_state(self) -> u32 {
+        bits_of(self.eax, 5, 7)
+    }
+
+    pub fn maximum_bytes_for_enabled_features(self) -> u32 {
+        self.ebx
+    }
+
+    pub fn maximum_bytes_for_supported_features(self) -> u32 {
+        self.ecx
+    }
+}
+
+impl fmt::Debug for ProcessorExtendedState {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        dump!(self, f, "ProcessorExtendedState", {
+            x87_state,
+            sse_state,
+            avx_state,
+            mpx_state,
+            avx_512_state,
+            ia32_xss,
+            pkru_state,
+
+            maximum_bytes_for_enabled_features,
+
+            maximum_bytes_for_supported_features
+        })
+    }
+}
+
+#[derive(Copy,Clone)]
+pub struct ProcessorExtendedStateSecondary {
+    eax: u32,
+    ebx: u32,
+    ecx: u32,
+}
+
+impl ProcessorExtendedStateSecondary {
+    fn new() -> ProcessorExtendedStateSecondary {
+        let (a, b, c, _) = cpuid_ext(RequestType::ProcessorExtendedState, 0x00000001);
+        ProcessorExtendedStateSecondary { eax: a, ebx: b, ecx: c }
+    }
+
+    bit!(eax, {
+        0 => xsaveopt_available,
```

Comment:
I'd just call these ones after their `CPUID` names in the Intel reference:

0: `xsaveopt`
1: `xsavec` (implies `xrstor` support for compact states; there is no `xrstorc` instruction)
2: `xgetbv`
3: `xsaves` (implies `xrstors`)

Reply:
I'm going to change these a little (not sure why I put `available`, that's kind of the point of CPUID), but I do like to have a wee bit more verbosity.